net/ip.h

```diff
@@ -0,0 +1,38 @@
+// SPDX-FileCopyrightText: 2023 David Zero <zero-one@zer0-one.net>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+#ifndef NET_IP_H_
+#define NET_IP_H_
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+namespace net {
+
+struct IP4Address {
+    IP4Address(uint32_t address) { addr = address; }
+
+    uint32_t addr;
+
+    std::string serialize(void);
+};
+
+struct IP6Address {
+    IP6Address(std::vector<uint16_t> address) {
```

Comment:
How about something like `std::span<std::uint16_t, 8>` or `std::array<std::uint16_t, 8>` instead? Neither would require heap allocations the way a std::vector does and it would also enforce from the call-site that you pass something with exactly 8 elements, so you don't have the silently doing the wrong thing and reading from uninitialized memory the current implementation has.